### PR TITLE
ARROW-304: NullableMapReaderImpl.isSet() always returns true

### PR DESCRIPTION
**Pull request by julienledem**

### CODE REVIEW

File: java/vector/src/test/java/org/apache/arrow/vector/complex/writer/TestComplexWriter.java [hunk 1]
```diff
@@ -77,28 +85,31 @@ public void nullableMap() {
...
     }
     writer.setValueCount(COUNT);
     MapReader rootReader = new SingleMapReaderImpl(parent).reader("root");
     for (int i = 0; i < COUNT; i++) {
       rootReader.setPosition(i);
+      FieldReader map = rootReader.reader("map");
       if (i % 2 == 0) {
-        Assert.assertNotNull(rootReader.reader("map").readObject());
-        Assert.assertEquals(i, rootReader.reader("map").reader("nested").readLong().longValue());
+        assertTrue("index is set: " + i, map.isSet());
```
Comment (adeneche) [inline java/vector/src/test/java/org/apache/arrow/vector/complex/writer/TestComplexWriter.java]: can you add an assertion to ensure ```rootReader.isSet()```

File: java/vector/src/test/java/org/apache/arrow/vector/complex/writer/TestComplexWriter.java [hunk 2]
```diff
@@ -77,28 +85,31 @@ public void nullableMap() {
...
+        assertNotNull("index is set: " + i, map.readObject());
+        assertEquals(i, map.reader("nested").readLong().longValue());
       } else {
-        Assert.assertNull(rootReader.reader("map").readObject());
+        assertFalse("index is not set: " + i, map.isSet());
```
Comment (adeneche) [inline java/vector/src/test/java/org/apache/arrow/vector/complex/writer/TestComplexWriter.java]: can you also add an assertion to ensure ```!rootReader.isSet()```

File: java/vector/src/test/java/org/apache/arrow/vector/complex/writer/TestComplexWriter.java [hunk 3]
```diff
@@ -77,28 +85,31 @@ public void nullableMap() {
     MapVector parent = new MapVector("parent", allocator, null);
     ComplexWriter writer = new ComplexWriterImpl("root", parent);
     MapWriter rootWriter = writer.rootAsMap();
-    MapWriter mapWriter = rootWriter.map("map");
-    BigIntWriter nested = mapWriter.bigInt("nested");
     for (int i = 0; i < COUNT; i++) {
+      rootWriter.start();
```
Comment (adeneche) [inline java/vector/src/test/java/org/apache/arrow/vector/complex/writer/TestComplexWriter.java]: don't we need to call ```setPosition()``` on rootWriter ?

File: java/vector/src/test/java/org/apache/arrow/vector/complex/writer/TestComplexWriter.java (hunk 1 above)
Comment (julienledem) [inline java/vector/src/test/java/org/apache/arrow/vector/complex/writer/TestComplexWriter.java]: will do.

File: java/vector/src/test/java/org/apache/arrow/vector/complex/writer/TestComplexWriter.java (hunk 2 above)
Comment (julienledem) [inline java/vector/src/test/java/org/apache/arrow/vector/complex/writer/TestComplexWriter.java]: unlike rootReader.reader("map") rootReader is always set. Adding the check

File: java/vector/src/test/java/org/apache/arrow/vector/complex/writer/TestComplexWriter.java (hunk 3 above)
Comment (julienledem) [inline java/vector/src/test/java/org/apache/arrow/vector/complex/writer/TestComplexWriter.java]: adding

**Review by adeneche — APPROVED** [review]
+1 LGTM

**Pull request closed without merging**